### ISSUE TITLE
os.getpid() not working

### ISSUE THREAD

**Issue by mikus** [scijava/jython-shaded#5]
When I try to invoke os.getpid() using jython-shaded-2.7.1.jar I get a following error:
```
$ /usr/lib/jvm/java-8-oracle/bin/java -cp jython-shaded-2.7.1.jar org.python.util.jython <<< "import os; os.getpid()"
Jython 2.7.1 (default:0df7adb1b397, Jun 30 2017, 19:02:43) 
[Java HotSpot(TM) 64-Bit Server VM (Oracle Corporation)] on java1.8.0_151
Type "help", "copyright", "credits" or "license" for more information.
>>> import os; os.getpid()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'getpid'
```
whereas using jython-standalone-2.7.1.jar it works perfectly
```
$ /usr/lib/jvm/java-8-oracle/bin/java -cp jython-standalone-2.7.1.jar org.python.util.jython <<< "import os; os.getpid()"
28106

```

**Comment by ctrueden** [comment]
The problem is:
```
Failed to load native POSIX impl; falling back on Java impl. Stacktrace follows.
java.lang.NoClassDefFoundError: Could not initialize class org.scijava.jython.shaded.jnr.posix.POSIXFactory$DefaultLibCProvider$SingletonHolder
       at org.scijava.jython.shaded.jnr.posix.POSIXFactory$DefaultLibCProvider.getLibC(POSIXFactory.java:283)
```

But I don't know how to fix it yet.

**Comment by ctrueden** [comment]
I thought maybe it would help to not shade the `jni/` folder. (Intuitively, I thought maybe the `jffi` native libraries e.g. `jni/x86_64-Linux/libjffi-1.2.so` were being loaded with absolute resource paths, so this might fix the problem—and there would be no clashes with other JFFI versions because these native libs have the version number in the filenames.) But unfortunately, removing that `<relocation>` stanza from the `deps/pom.xml` has no effect on this error.

Next step is probably to step in with a debugger to see what happens under the hood when an `os` function is called unsuccessfully, and maybe also do the same when the same function is called successfully with `jython-standalone`. See where & how the code paths diverge. But I may not have time to pursue that for a few weeks. If someone else can try in the meantime, it would be much appreciated!

**Comment by ctrueden** [comment]
As a starting point: the following program returns `0` always:
```java
POSIX posix = org.scijava.jython.shaded.jnr.posix.POSIXFactory.getPOSIX();
int pid = posix.getpid();
System.out.println(pid);
```

Whereas using `jnr.posix.POSIXFactory.getPOSIX();` instead returns a non-zero PID.

I started debugging into it in Eclipse, and it seems clear to me that it is necessary, but not sufficient, for the JFFI `jni` folder to live off the root. At least, the stock JFFI loads it successfully on my systems, whereas the embedded/shaded JFFI fails.

I wish I had more time to dig in to this but I really have to stop now, and will not have time again till the latter half of December.

**Comment by larsrh** [comment]
Thanks for the detective work so far, @ctrueden. I'll try to reproduce your example in my system, maybe I can get a hold of the problem.

**Comment by headius** [comment]
Commenting so I'll get notifications. If there's anything we should fix in jnr-* please let me know and file an issue.

I wonder if something in jffi is using getClass.getPackageName for something, and getting screwed up because it has changed? 

We want to start vendoring everything in JRuby so this is relevant for our work as well.

**Comment by larsrh** [comment]
I think I'm getting closer to the actual problem.

I have two projects differing in dependencies (please forgive the Scala/sbt syntax):

    libraryDependencies += "org.scijava" % "jython-shaded" % "2.7.1"

    libraryDependencies += "com.github.jnr" % "jnr-posix" % "3.0.41"

3.0.41 is the version of `jnr-posix` as specified in the [Jython build](https://github.com/jythontools/jython/blob/v2.7.1/build.xml#L161).

Here is some code to disambiguate these two dependencies:

```scala
/* import either of these */
// import jnr.posix.POSIXFactory
// import org.scijava.jython.shaded.jnr.posix.POSIXFactory

val posix = POSIXFactory.getPOSIX

val meth = posix.getClass.getDeclaredMethod("posix")
meth.setAccessible(true)

val actualPosix = meth.invoke(posix)
```

In the non-shaded case, the result looks like this:

```
posix: jnr.posix.POSIX = jnr.posix.LazyPOSIX@b9393b5

meth: java.lang.reflect.Method = private final jnr.posix.POSIX jnr.posix.LazyPOSIX.posix()
res0: Unit = ()

actualPosix: Object = jnr.posix.CheckedPOSIX@254223c0
```

In the shaded case:

```
posix: org.scijava.jython.shaded.jnr.posix.POSIX = org.scijava.jython.shaded.jnr.posix.LazyPOSIX@528e4df0

meth: java.lang.reflect.Method = private final org.scijava.jython.shaded.jnr.posix.POSIX org.scijava.jython.shaded.jnr.posix.LazyPOSIX.posix()
res0: Unit = ()

actualPosix: Object = org.scijava.jython.shaded.jnr.posix.JavaPOSIX@65b73714
```

Now just to figure out where the lazy loading code diverges.

**Comment by larsrh** [comment]
I had a hunch, and I just confirmed that hunch: it's an `UnsatisfiedLinkError` that gets swallowed. Here's the relevant code in `POSIXFactory`:

```java
try {
    posix = loadNativePOSIX(handler);
    posix = posix != null ? new CheckedPOSIX(posix, handler) : null;
    // ENEBO: Should printing be done through a handler+log method?
    if (handler.isVerbose()) {
        if (posix != null) {
            System.err.println("Successfully loaded native POSIX impl.");
        } else {
            System.err.println("Failed to load native POSIX impl; falling back on Java impl. Unsupported OS.");
        }
    }
} catch (Throwable t) {
    if (handler.isVerbose()) {
        System.err.println("Failed to load native POSIX impl; falling back on Java impl. Stacktrace follows.");
        t.printStackTrace();
    }
}
```

The error gets thrown in the constructor of (on my system) `LinuxPOSIX`, more precisely, when executing the static initializer of `POSIXFactory.DefaultLibCProvider.SingletonHolder`.

I have currently narrowed it down to the method `Library.getLibrary`: In the non-shaded case, the first line returns a `NativeLibraryLoader`, in the shaded case something invalid. My knowledge about native bindings is very limited, though.

**Comment by larsrh** [comment]
OK, I think I have an idea of what the problem actually is. To put it plain and simple: the package name is part of the C interface. (@headius surely knows this, it's just that I had to make this discovery the hard way …) So, shading `com.kenai.jffi.Foreign` to something else means that suddenly the expected C function names are just different, hence it is unable to call anything.

**Comment by ctrueden** [comment]
@larsrh Thanks for all the follow-up work, and very sorry for the long delay in reply on my side.

I finally did a little more investigation into this issue!

Firstly, I confirm that `os.getpid()` works if the `jffi-1.2.17` and `jffi-1.2.17-native` libraries (i.e.: paths beginning with `com/kenai/jffi` and `jni`) are left unshaded:

<details>

```diff
$ git diff
diff --git a/deps/pom.xml b/deps/pom.xml
index 01426df..b718eac 100644
--- a/deps/pom.xml
+++ b/deps/pom.xml
@@ -59,8 +59,8 @@
                                                        <shadedPattern>org.scijava.jython.shaded.com.google</shadedPattern>
                                                </relocation>
                                                <relocation>
-                                                       <pattern>com.kenai</pattern>
-                                                       <shadedPattern>org.scijava.jython.shaded.com.kenai</shadedPattern>
+                                                       <pattern>com.kenai.constantine</pattern>
+                                                       <shadedPattern>org.scijava.jython.shaded.com.kenai.constantine</shadedPattern>
                                                </relocation>
                                                <relocation>
                                                        <pattern>com.xhaus</pattern>
@@ -82,10 +82,6 @@
                                                        <pattern>jline</pattern>
                                                        <shadedPattern>org.scijava.jython.shaded.jline</shadedPattern>
                                                </relocation>
-                                               <relocation>
-                                                       <pattern>jni</pattern>
-                                                       <shadedPattern>org.scijava.jython.shaded.jni</shadedPattern>
-                                               </relocation>
                                                <relocation>
                                                        <pattern>jnr</pattern>
                                                        <shadedPattern>org.scijava.jython.shaded.jnr</shadedPattern>
$ jar tf shaded/target/jython-shaded-2.7.2-SNAPSHOT.jar | grep -v '^\(org/scijava\|Lib\|org/python\|META-INF\)' | grep '/$'
com/
com/kenai/
com/kenai/jffi/
com/kenai/jffi/internal/
org/
jni/
jni/Darwin/
jni/aarch64-Linux/
jni/arm-Linux/
jni/i386-Linux/
jni/i386-SunOS/
jni/i386-Windows/
jni/ppc-AIX/
jni/ppc64-Linux/
jni/ppc64le-Linux/
jni/sparcv9-SunOS/
jni/x86_64-FreeBSD/
jni/x86_64-Linux/
jni/x86_64-OpenBSD/
jni/x86_64-SunOS/
jni/x86_64-Windows/
$ java -jar shaded/target/jython-shaded-2.7.2-SNAPSHOT.jar -c 'import os; print(os.getpid())'
25407
```

</details>

I pushed 295ae9692441c5dba35c5059e615952a4514a2b0 implementing this change.

However, what I think would be nicer would be to exclude the JFFI files from the jython-shaded artifact altogether, in favor of declaring `jffi-1.2.17` and `jffi-1.2.17-native` as transitive dependencies. I think this would be OK, because both Jython and JRuby appear to use 1.2.17. (Technically, Jython uses `jffi-1.2.17` and `jffi-1.2.16-native`, which is weird, but I really hope this is merely sloppiness rather than an intentional workaround for something!).

Investigating how to wrangle the shade plugin to do that now.

**Comment by ctrueden** [comment]
I figured it out! `\^_^/` ffa3b7718d8aabc914fb78a5180ceb20649c69bc

So now, `jython-shaded` has two transitive dependencies, `jffi-1.2.17` and `jffi-1.2.17-native`, and the entries from those artifacts are excluded from `jython-shaded` completely. This lets functions like `os.getpid()` function as desired, as long as the JFFI libraries are on the classpath.

<details>

```
$ jar tf shaded/target/jython-shaded-2.7.2-SNAPSHOT.jar | grep jffi|grep -v '^\(org/scijava\|org\/python\)' | wc -l
       0
$ grep -A 14 '<dependencies>' shaded/dependency-reduced-pom.xml
  <dependencies>
    <dependency>
      <groupId>com.github.jnr</groupId>
      <artifactId>jffi</artifactId>
      <version>1.2.17</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>com.github.jnr</groupId>
      <artifactId>jffi</artifactId>
      <version>1.2.17</version>
      <classifier>native</classifier>
      <scope>compile</scope>
    </dependency>
  </dependencies>
$ java -cp $HOME/.m2/repository/com/github/jnr/jffi/1.2.17/jffi-1.2.17.jar:$HOME/.m2/repository/com/github/jnr/jffi/1.2.17/jffi-1.2.17-native.jar:shaded/target/jython-shaded-2.7.2-SNAPSHOT.jar org.python.util.jython -c 'import os; print(os.getpid())'
28602
```

And as expected, it doesn't work if `jffi-1.2.17` and/or `jffi-1.2.17-native` is missing from the classpath:

```
$ java -cp shaded/target/jython-shaded-2.7.2-SNAPSHOT.jar org.python.util.jython -c 'import os; print(os.getpid())'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'module' object has no attribute 'getpid'
```

</details>

@larsrh @mikus Please give the latest master branch a try, and let me know how it goes for you. If you report back that it works, I will release jython-shaded 2.7.1.1 to OSS Sonatype. Or if it still fails, please reopen the issue and we can troubleshoot further.

**Comment by larsrh** [comment]
@ctrueden I just gave it a spin, but for some weird reason, I can't even reproduce the original error in the context of my code.

**Comment by larsrh** [comment]
However, I just tried building this from the repository:

```
$ java -jar shaded/target/jython-shaded-2.7.2-SNAPSHOT.jar 
Jython 2.7.1 (default:0df7adb1b397, Jun 30 2017, 19:02:43) 
[OpenJDK 64-Bit Server VM (Oracle Corporation)] on java1.8.0_181
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.getpid()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'getpid'
```

**Comment by ctrueden** [comment]
@larsrh Sorry for the long delay in reply. As I described above: it is now the case that `jython-shaded` depends on the `jffi` main and native JARs. Since you ran with `java -jar`, those transitive dependencies are not on your classpath, so things do not work.

It should work if you include `jffi-1.2.17.jar` and `jffi-1.2.17-native.jar` on your classpath too:
```
$ mvn dependency:copy-dependencies
...
[INFO] Copying jffi-1.2.17.jar to .../jython-shaded/shaded/target/dependency/jffi-1.2.17.jar
[INFO] Copying jffi-1.2.17-native.jar to .../jython-shaded/shaded/target/dependency/jffi-1.2.17-native.jar
...
$ java -cp 'shaded/target/jython-shaded-2.7.2-SNAPSHOT.jar:shaded/target/dependency/jffi-1.2.17.jar:shaded/target/dependency/jffi-1.2.17-native.jar' org.python.util.jython -c 'import os; print(os.getpid())'
42403
```

**Comment by ctrueden** [comment]
I have now released jython-shaded version 2.7.1.1 to OSS Sonatype.

Tested with [jgo](https://github.com/scijava/jgo) and it works:

```
$ jgo org.scijava:jython-shaded:2.7.1.1
Jython 2.7.1 (default:0df7adb1b397, Jun 30 2017, 19:02:43)
[Java HotSpot(TM) 64-Bit Server VM (Oracle Corporation)] on java1.8.0_172
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> print(os.getpid())
44896
```

**Comment by larsrh** [comment]
Great, thanks!
